src/pybind/mgr/mds_autoscaler/module.py

```diff
@@ -0,0 +1,87 @@
+"""
+Automatically scale MDSs based on status of the file-system using the FSMap
+"""
+
+import errno
+import logging
+from mgr_module import MgrModule
+
+import orchestrator
+
+log = logging.getLogger(__name__)
+
+"""
+Some terminology is made up for the purposes of this module:
+
+ - "grow/shrink" - increase/decrease the MDSs in the system
+"""
+
+
+class MDSAutoscaler(MgrModule):
+    """
+    MDS autoscaler.
+    """
+    def __init__(self, *args, **kwargs):
+        super(MDSAutoscaler, self).__init__(*args, **kwargs)
+        self.fs_map = None
+
+    def create_mds(self, fs_name):
+        spec = orchestrator.ServiceSpec(fs_name)
+        try:
+            completion = self.add_mds(spec)
+            self._orchestrator_wait([completion])
+            orchestrator.raise_if_exception(completion)
+        except Exception as e:
+            # Don't let detailed orchestrator exceptions (python backtraces)
+            # bubble out to the user
+            self.log.exception("Failed to create MDS daemon")
+            return -errno.EINVAL, "", str(e)
+        return 0, "", ""
+
+    def delete_mds(self):
+        name = None
+        try:
+            for sb in self.fs_map['standbys']:
+                # we attempt to delete only idle standby process
+                if sb['state'] == 'up:active' and sb['rank'] == -1:
+                    name = sb['name']
+                    self.remove_mds(sb['name'])
+                    break
```

Comment:
I think this logic should be moved to cephadm so that it, in general, kills standby (mds,mgr) daemons when scaling down.

Reply:
@liewegas I presumed that cephadm was only responsible for implementing the mechanism of the operations and the policy decisions were supposed to be taken by cephadm clients.

Reply:
cephadm has a kubernetes-like spec (that says, e.g., "3 daemons") and controller that ensures that the right number of daemons are running, removing or adding individual daemons will tend to fight with that: if you remove one, a new one will get created (somewhere), and if you add one, an existing one will get removed.  Alternatively, if you adjust the count down, cephadm will pick which one to remove.

I think the best path forward is to make cephadm (and rook) just a bit smarter so that they prefer to remove standby daemons and not active ones.  Then this tool just needs to adjust the total daemon count (max_mds + num standbys)...

Reply:
I have a half-written patch to make cephadm prefer to remove standbys... so I suggest simplifying this PR to just adjust the count.

Reply:
@liewegas "Then this tool just needs to adjust the total daemon count (max_mds + num standbys)" ... how do I do that ?

Reply:
In the most common case, it's as simple as ``ceph orch apply mds $fsname $count`` (apply_mds(...)).  But the user might have a customized placement (specifying a label, hosts, etc.).  So I think right thing to do is fetch the current placement (you can do this with ``ceph orch ls --service-name mds.$fsname`` (describe_service(service_name='mds.whatever')), modify the spec.placement PlacementSpec (make a copy first--don't modify in place), and then send that back to apply_mds().

Reply:
@liewegas all that the `describe_service()` returns back is one record about the `crash` service
I've posted a WIP PR; please take a look; I must be missing something that I can't fathom yet.

Reply:
@sebastian-philipp I'm trying to test my plugin on a vstart cluster on my laptop. I can't find anyway to make cephadm act on the default cephfs.a cluster. Looks like cephadm should start the daemons itself to gain ownership of the cluster to effectively act on them.
Is there a way to instruct cephadm to start a file-system and take ownership of all relevant daemons ?
The MDS Autoscaler Plugin is supposed to act on changes to the FSMap and spawn or kill MDS standby daemons as required. Could you suggest me a workflow to achieve this ? I've used Sage's suggestions to update my code but the describe_service() API isn't listing anything than the crash service. Same goes with `ceph orch ls` command at the command-line as well.

Reply:
@mchangir so, cephadm won't touch any daemons that are not created by cephadm, like e.g. daemons created by vstart.sh. you'll have to create them using `ceph orch apply mds...` and not via vstart.

> is there a way to instruct cephadm to start a file-system and take ownership of all relevant daemons ?

yes: `ceph fs volume create --placement=...`.

>  the describe_service() API isn't listing anything than the crash service

exactly. you'll have to create the daemons using cephadm first.